// eclg-prospero/test/integration/subscription.test.js
var assert = require('assert');
var querystring = require('querystring');
var http = require('http');
var Prospero = require('../../lib/prospero');

var sharedConfig = {
    rootUrl: 'http://localhost:4778',
    principal: "ONE",
    sharedKey: "1234567890123456",
    prosperoDefaults: {}
};



exports['test subscribe, get and unsubscribe'] = function(){
    var obj = {
        messageType: "Node.Prospero.Test2",
        callbackUrl: "http://mycallbackurl.com/a/me",
        tags: { UserID: "joe" }  
    };


    var p = new Prospero(sharedConfig);
    p.subscribe(obj, function(error, data) {
        assert.ok(!error);
        assert.ok(data.data.subscription.id);
        assert.equal(data.statusCode, 200);
        var firstSubId = data.data.subscription.id;

        // try to subscribe again
        p.subscribe(obj, function(error, data) {
            assert.ok(!error);
            assert.equal(data.statusCode, 200);
            assert.equal(firstSubId, data.data.subscription.id);
                console.log(require('util').inspect(data, true, 10));

            p.unsubscribe(firstSubId, function(error, data) {
                assert.ok(!error);
                assert.equal(200, data.statusCode)

                p.getSubscription(firstSubId, function(error, data) {
                    assert.ok(!error);
                    assert.equal(404, data.statusCode)
                })
            });
        });
    });    
};
